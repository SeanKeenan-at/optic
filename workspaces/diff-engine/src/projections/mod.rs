pub mod conflicts;
pub mod endpoint;
pub mod history;
pub mod shape;
pub mod spec_events;

pub use conflicts::ConflictsProjection;
pub use endpoint::EndpointProjection;
pub use history::{CommitId, HistoryProjection};
pub use shape::ShapeProjection;
pub use spec_events::{SpecAssemblerError, SpecAssemblerProjection};

use crate::events::{EndpointEvent, RfcEvent, ShapeEvent, SpecEvent};
use cqrs_core::{Aggregate, AggregateCommand, AggregateEvent, CommandError};
use std::error::Error;

#[derive(Debug)]
pub struct SpecProjection {
  endpoint: endpoint::EndpointProjection,
  history: history::HistoryProjection,
  shape: shape::ShapeProjection,
  conflicts: conflicts::ConflictsProjection,
}

impl Default for SpecProjection {
  fn default() -> Self {
    Self {
      endpoint: EndpointProjection::default(),
      history: HistoryProjection::default(),
      shape: ShapeProjection::default(),
      conflicts: ConflictsProjection::default(),
    }
  }
}

impl SpecProjection {
  pub fn endpoint(&self) -> &EndpointProjection {
    &self.endpoint
  }

  pub fn history(&self) -> &HistoryProjection {
    &self.history
  }

  pub fn shape(&self) -> &ShapeProjection {
    &self.shape
  }
  pub fn conflicts(&self) -> &ConflictsProjection {
    &self.conflicts
  }
}

impl Aggregate for SpecProjection {
  fn aggregate_type() -> &'static str {
    "spec_projection"
  }
}

// Events
// ------

impl AggregateEvent<SpecProjection> for SpecEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    match self {
      SpecEvent::EndpointEvent(event) => {
        projection.endpoint.apply(event.clone());
        projection.conflicts.apply(event);
      }
      SpecEvent::ShapeEvent(event) => {
        projection.shape.apply(event.clone());
        projection.conflicts.apply(event);
      }
      SpecEvent::RfcEvent(event) => projection.history.apply(event),
    }
  }
}

impl AggregateEvent<HistoryProjection> for SpecEvent {
  fn apply_to(self, projection: &mut HistoryProjection) {
    if let SpecEvent::RfcEvent(event) = self {
      event.apply_to(projection);
    }
  }
}

impl<I> From<I> for SpecProjection
where
  I: IntoIterator,
  I::Item: AggregateEvent<Self>,
{
  fn from(events: I) -> Self {
    let mut projection = SpecProjection::default();
    for event in events.into_iter() {
      projection.apply(event);
    }
    projection
  }
}

// Convient application of spec event variants
// -------------------------------------------
impl AggregateEvent<SpecProjection> for ShapeEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    projection.apply(SpecEvent::ShapeEvent(self))
  }
}

impl AggregateEvent<SpecProjection> for RfcEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    projection.apply(SpecEvent::RfcEvent(self))
  }
}

impl AggregateEvent<SpecProjection> for EndpointEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    projection.apply(SpecEvent::EndpointEvent(self))
  }
}
