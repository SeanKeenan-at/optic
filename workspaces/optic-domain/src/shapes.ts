// Shape events
export enum ICoreShapeKinds {
  ObjectKind = '$object',
  ListKind = '$list',
  MapKind = '$map',
  OneOfKind = '$oneOf',
  AnyKind = '$any',
  StringKind = '$string',
  NumberKind = '$number',
  BooleanKind = '$boolean',
  NullableKind = '$nullable',
  OptionalKind = '$optional',
  UnknownKind = '$unknown',
}

export enum ICoreShapeInnerParameterNames {
  ListInner = '$listItem',
  NullableInner = '$nullableInner',
  OptionalInner = '$optionalInner',
}

// Optic engine shape types
export enum JsonLike {
  OBJECT = 'Object',
  ARRAY = 'Array',
  NULL = 'Null',
  STRING = 'String',
  NUMBER = 'Number',
  BOOLEAN = 'Boolean',
  UNDEFINED = 'Undefined',
}
