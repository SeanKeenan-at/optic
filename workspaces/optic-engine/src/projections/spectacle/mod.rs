pub mod endpoints;
pub mod shape;
