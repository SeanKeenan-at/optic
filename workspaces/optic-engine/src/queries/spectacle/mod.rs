pub mod spec_choices;

pub use spec_choices::*;
