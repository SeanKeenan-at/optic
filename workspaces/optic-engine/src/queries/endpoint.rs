use crate::commands::{EndpointCommand, SpecCommand};
use crate::events::HttpInteraction;
use crate::projections::endpoint::{Edge, EndpointProjection, Node, ROOT_PATH_ID};
use crate::projections::endpoint::{RequestBodyDescriptor, ResponseBodyDescriptor};
use crate::state::endpoint::{
  HttpMethod, HttpStatusCode, PathComponentId, PathComponentIdRef, RequestId, ResponseId,
};
use petgraph::graph::Graph;
use petgraph::visit::EdgeFilteredNeighborsDirected;
use serde::Serialize;

pub struct EndpointQueries<'a> {
  pub endpoint_projection: &'a EndpointProjection,
}

impl<'a> EndpointQueries<'a> {
  pub fn new(endpoint_projection: &'a EndpointProjection) -> Self {
    EndpointQueries {
      endpoint_projection,
    }
  }

  fn extract_normalized_path(path: &str) -> &str {
    if path.eq("/") {
      path
    } else if path.ends_with("/") {
      &path[..path.len() - 1]
    } else {
      path
    }
  }

  pub fn resolve_interaction_path(
    &self,
    interaction: &HttpInteraction,
  ) -> Option<PathComponentIdRef> {
    self.resolve_path(&interaction.request.path)
  }

  pub fn resolve_path(&self, path: &str) -> Option<PathComponentIdRef> {
    let path = Self::extract_normalized_path(path);
    // eprintln!("{}", path);
    let mut path_components = path.split('/');
    // skip leading empty
    path_components.next();
    let mut last_resolved_path_id = Some(ROOT_PATH_ID);
    while let Some(s) = path_components.next() {
      // eprintln!("trying to match segment {}", s);
      let node_index = self
        .graph_get_index(last_resolved_path_id.unwrap())
        .expect("expected a node with node_id to exist");

      last_resolved_path_id = None;

      // first search for named path components
      let children = self.graph_get_children(node_index);
      for child in children {
        let child_node = self.endpoint_projection.graph.node_weight(child).unwrap();

        match child_node {
          Node::PathComponent(id, descriptor) => {
            // eprintln!("1 - neighbor {}", descriptor.name);
            if !descriptor.is_parameter {
              if descriptor.name == s {
                last_resolved_path_id = Some(id);
                break;
              }
            }
          }
          _ => {}
        }
      }
      if let Some(x) = last_resolved_path_id {
        continue;
      }
      // try path parameters since we didn't find any matches in regular path components
      let children = self.graph_get_children(node_index);
      for child in children {
        let child_node = self.endpoint_projection.graph.node_weight(child).unwrap();
        match child_node {
          Node::PathComponent(id, descriptor) => {
            // eprintln!("2 - neighbor {}", descriptor.name);

            if descriptor.is_parameter {
              last_resolved_path_id = Some(id);
              break;
            }
          }
          _ => {}
        }
      }
      if let Some(x) = last_resolved_path_id {
        continue;
      }
      return None;
    }
    last_resolved_path_id
  }

  pub fn resolve_operations_by_request_method(
    &self,
    method: &'a String,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)> {
    self
      .resolve_requests(path_id, method)
      .expect("expected a operations to exist")
  }

  pub fn resolve_operations(
    &self,
    interaction: &'a HttpInteraction,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)> {
    self.resolve_operations_by_request_method(&interaction.request.method, path_id)
  }

  pub fn resolve_requests(
    &self,
    path_id: PathComponentIdRef,
    method: &'a String,
  ) -> Option<impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)>> {
    let path_node_index = self.graph_get_index(path_id)?;
    let children = self
      .endpoint_projection
      .graph
      .neighbors_directed(*path_node_index, petgraph::Direction::Incoming);
    let matching_method = children
      .filter(move |i| {
        let node = self.endpoint_projection.graph.node_weight(*i).unwrap();
        match node {
          Node::HttpMethod(http_method) => method == http_method,
          _ => false,
        }
      })
      .flat_map(move |i| {
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);

        let operations = children.filter_map(move |i| {
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          match node {
            Node::Request(request_id, request_descriptor) => Some((request_id, request_descriptor)),
            _ => None,
          }
        });
        operations
      });
    Some(matching_method)
  }

  pub fn resolve_request_by_method_and_content_type(
    &self,
    path_id: PathComponentIdRef,
    method: &'a String,
    content_type: Option<&'a String>,
  ) -> Option<(&RequestId, &RequestBodyDescriptor)> {
    self.resolve_requests(path_id, method).and_then(|mut it| {
      it.find(|(id, body)| match content_type {
        Some(content_type) => match body.body {
          Some(ref body) => body.http_content_type.eq(content_type),
          None => false,
        },
        None => body.body.is_none(),
      })
    })
  }

  pub fn resolve_responses(
    &self,
    path_id: &'a PathComponentId,
    method: &'a String,
  ) -> Option<impl Iterator<Item = (&ResponseId, &ResponseBodyDescriptor)>> {
    let response_nodes = self
      .endpoint_projection
      .get_response_nodes(path_id, method)?;

    Some(response_nodes.map(|node| match node {
      Node::Response(response_id, body_descriptor) => (response_id, body_descriptor),
      _ => unreachable!("get response nodes should only return response nodes"),
    }))
  }

  pub fn resolve_response_by_method_status_code_and_content_type(
    &self,
    path_id: PathComponentIdRef,
    method: &'a str,
    status_code: u16,
    content_type: Option<&'a String>,
  ) -> Option<(&ResponseId, &ResponseBodyDescriptor)> {
    self
      .resolve_responses_by_method_and_status_code(method, status_code, path_id)
      .find(|(id, body)| match content_type {
        Some(content_type) => match body.body {
          Some(ref body) => body.http_content_type.eq(content_type),
          None => false,
        },
        None => body.body.is_none(),
      })
  }

  pub fn resolve_responses_by_method_and_status_code(
    &self,
    method: &'a str,
    status_code: u16,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&ResponseId, &ResponseBodyDescriptor)> {
    let path_node_index = self
      .graph_get_index(path_id)
      .expect("expected a node with node_id to exist");
    let children = self
      .endpoint_projection
      .graph
      .neighbors_directed(*path_node_index, petgraph::Direction::Incoming);
    let matching_status_code = children
      .filter(move |i| {
        let node = self.endpoint_projection.graph.node_weight(*i).unwrap();
        // eprintln!("method node {:?}", node);
        match node {
          Node::HttpMethod(http_method) => method == *http_method,
          _ => false,
        }
      })
      .flat_map(move |i| {
        // eprintln!("method node id {:?}", i);
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);
        let status_code_nodes = children.filter_map(move |i| {
          // eprintln!("method child node id {:?}", i);
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          match node {
            Node::HttpStatusCode(http_status_code) => {
              // eprintln!("status code {:?}", http_status_code);
              if status_code == *http_status_code {
                Some(i)
              } else {
                None
              }
            }
            _ => None,
          }
        });
        status_code_nodes
      })
      .flat_map(move |i| {
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);
        let response_nodes = children.filter_map(move |i| {
          // eprintln!("status_code child node id {:?}", i);
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          // eprintln!("status_code child node {:?}", node);
          match node {
            Node::Response(response_id, response_descriptor) => {
              Some((response_id, response_descriptor))
            }
            _ => None,
          }
        });
        response_nodes
      });
    matching_status_code
  }

  pub fn delete_endpoint_commands(
    &self,
    path_id: &'a PathComponentId,
    method: &'a HttpMethod,
  ) -> Option<DeleteEndpointCommands> {
    let request_ids = self
      .resolve_requests(path_id, method)?
      .map(|(request_id, _)| request_id);
    let response_ids = self
      .resolve_responses(path_id, method)?
      .map(|(response_id, _)| response_id);

    let request_commands = request_ids.cloned().map(EndpointCommand::remove_request);
    let response_commands = response_ids.cloned().map(EndpointCommand::remove_response);

    Some(DeleteEndpointCommands {
      path_id: path_id.clone(),
      method: method.clone(),
      commands: request_commands
        .chain(response_commands)
        .map(SpecCommand::from)
        .collect(),
    })
  }

  fn graph_get_index(&self, node_id: &str) -> Option<&petgraph::graph::NodeIndex> {
    self.endpoint_projection.node_id_to_index.get(node_id)
  }

  fn graph_get_node(&self, node_id: &str) -> &Node {
    self
      .endpoint_projection
      .graph
      .node_weight(
        *self
          .graph_get_index(node_id)
          .expect("expected a node with node_id to exist"),
      )
      .expect("expected node with node_id to exist")
  }

  fn graph_get_children(
    &self,
    node_index: &petgraph::graph::NodeIndex,
  ) -> petgraph::graph::Neighbors<Edge> {
    let neighbors = self
      .endpoint_projection
      .graph
      .neighbors_directed(*node_index, petgraph::Direction::Incoming);
    return neighbors;
  }
}

#[derive(Debug, Serialize)]
pub struct DeleteEndpointCommands {
  path_id: PathComponentId,
  method: HttpMethod,
  commands: Vec<SpecCommand>,
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::events::SpecEvent;
  use crate::projections::SpecProjection;
  use crate::Aggregate;
  use petgraph::dot::Dot;
  use serde_json::json;

  fn interaction_with_path(path: String) -> HttpInteraction {
    serde_json::from_value(json!(
      {
        "uuid": "id",
        "request": {
          "host": "example.com",
          "method": "GET",
          "path": path,
          "query": {
            "shapeHashV1Base64": null,
            "asJsonString": null,
            "asText": null
          },
          "headers": {
            "shapeHashV1Base64": null,
            "asJsonString": null,
            "asText": null
          },
          "body": {
            "contentType": null,
            "value": {
              "shapeHashV1Base64": null,
              "asJsonString": null,
              "asText": null
            }
          }
        },
        "response": {
          "statusCode": 200,
          "headers": {
            "shapeHashV1Base64": null,
            "asJsonString": null,
            "asText": null
          },
          "body": {
            "contentType": null,
            "value": {
              "shapeHashV1Base64": null,
              "asJsonString": null,
              "asText": null
            }
          }
        },
        "tags": []
      }
    ))
    .unwrap()
  }
  #[test]
  pub fn can_ignore_trailing_slash() {
    let interaction: HttpInteraction = interaction_with_path(String::from("/a/b/c/"));
    let normalized_path = EndpointQueries::extract_normalized_path(&interaction.request.path);
    assert_eq!(normalized_path, "/a/b/c")
  }

  #[test]
  pub fn can_handle_no_trailing_slash() {
    let interaction: HttpInteraction = interaction_with_path(String::from("/a/b/c"));
    let normalized_path = EndpointQueries::extract_normalized_path(&interaction.request.path);
    assert_eq!(normalized_path, "/a/b/c")
  }
  #[test]
  pub fn can_handle_root_path() {
    let interaction: HttpInteraction = interaction_with_path(String::from("/"));
    let normalized_path = EndpointQueries::extract_normalized_path(&interaction.request.path);
    assert_eq!(normalized_path, "/")
  }

  #[test]
  pub fn can_generate_delete_commands() {
    let events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": { "pathId": "path_1", "parentPathId": "root", "name": "posts" }},
      {"PathComponentAdded": { "pathId": "path_2", "parentPathId": "path_1", "name": "favourites" }},
      {"PathComponentAdded": { "pathId": "path_3", "parentPathId": "root", "name": "authors" }},

      {"RequestAdded": { "requestId": "request_1", "pathId": "path_2", "httpMethod": "GET"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_1", "pathId": "path_2", "httpMethod": "GET", "httpStatusCode": 200 }},

      {"RequestAdded": { "requestId": "request_2", "pathId": "path_2", "httpMethod": "POST"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_2", "pathId": "path_2", "httpMethod": "POST", "httpStatusCode": 201 }},

      {"RequestAdded": { "requestId": "request_3", "pathId": "path_3", "httpMethod": "GET"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_3", "pathId": "path_3", "httpMethod": "GET", "httpStatusCode": 200 }},
    ]))
    .expect("should be able to deserialize test events");

    let spec_projection = SpecProjection::from(events);
    // dbg!(Dot::with_config(&spec_projection.endpoint().graph, &[]));

    let endpoint_queries = EndpointQueries::new(spec_projection.endpoint());

    let subject_path = String::from("path_2");
    let subject_method = String::from("GET");
    let deleted_endpoint_commmands = endpoint_queries
      .delete_endpoint_commands(&subject_path, &subject_method)
      .expect("delete commands are generated for existing path and method");

    let updated_spec =
      assert_valid_commands(spec_projection.clone(), deleted_endpoint_commmands.commands);
    let updated_queries = EndpointQueries::new(&updated_spec.endpoint());
    let remaining_requests = updated_queries
      .resolve_requests(&subject_path, &subject_method)
      .unwrap()
      .collect::<Vec<_>>();
    let remaining_responses = updated_queries
      .resolve_responses(&subject_path, &subject_method)
      .unwrap()
      .collect::<Vec<_>>();

    // TODO: enable these assertions as the EndpointProjection handles the resulting events
    // assert_eq!(remaining_requests.len(), 0);
    // assert_eq!(remaining_responses.len(), 0);
  }

  fn assert_valid_commands(
    mut spec_projection: SpecProjection,
    commands: impl IntoIterator<Item = SpecCommand>,
  ) -> SpecProjection {
    // let mut spec_projection = SpecProjection::default();
    for command in commands {
      let events = spec_projection
        .execute(command)
        .expect("generated commands must be valid");

      for event in events {
        spec_projection.apply(event)
      }
    }

    spec_projection
  }
}
