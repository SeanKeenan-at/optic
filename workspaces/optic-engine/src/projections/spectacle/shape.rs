use crate::projections::shape::{Edge, Node, ShapeProjection};
use crate::queries::spectacle::spec_choices::{ShapeChoice, ShapeChoiceQueries};
use crate::shapes::ShapeTrail;
use crate::state::shape::ShapeId;
use serde::Serialize;
use std::collections::BTreeMap;
use std::iter::FromIterator;

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GraphNodesAndEdges<N, E> {
  nodes: Vec<N>,
  edges: Vec<(usize, usize, E)>,
  node_index_to_id: BTreeMap<String, String>,
}

pub type ShapeChoiceMapping = BTreeMap<ShapeId, Vec<ShapeChoice>>;
impl From<&ShapeProjection> for ShapeChoiceMapping {
  fn from(shape_projection: &ShapeProjection) -> Self {
    let choice_queries = ShapeChoiceQueries::from(shape_projection);
    let shape_choices = shape_projection
      .graph
      .node_indices()
      .filter_map(|i| {
        let node = shape_projection
          .graph
          .node_weight(i)
          .expect("node should exist");
        match node {
          Node::Shape(shape_node) => Some(shape_node),
          _ => None,
        }
      })
      .map(|shape_node| {
        let trail = ShapeTrail::new(shape_node.shape_id.clone());
        let choices = choice_queries.trail_choices(&trail).collect();
        (shape_node.shape_id.clone(), choices)
      });
    BTreeMap::from_iter(shape_choices)
  }
}

// duplicated in [projections/spectacle/endpoints.rs]
pub type SerializableGraph = GraphNodesAndEdges<Node, Edge>;
impl From<ShapeProjection> for SerializableGraph {
  fn from(shape_projection: ShapeProjection) -> Self {
    let (graph_nodes, graph_edges) = shape_projection.graph.into_nodes_edges();
    let nodes = graph_nodes.into_iter().map(|x| x.weight).collect();
    let edges = graph_edges
      .into_iter()
      .map(|x| (x.source().index(), x.target().index(), x.weight))
      .collect();
    let value: GraphNodesAndEdges<Node, Edge> = GraphNodesAndEdges {
      nodes,
      edges,
      node_index_to_id: BTreeMap::from_iter(
        shape_projection
          .node_id_to_index
          .into_iter()
          .map(|(k, v)| (v.index().to_string(), k)),
      ),
    };
    value
  }
}

impl ShapeProjection {
  pub fn to_json_string(&self) -> String {
    serde_json::to_string(&self.to_serializable_graph()).expect("graph should serialize")
  }

  pub fn to_serializable_graph(&self) -> SerializableGraph {
    let copy = self.clone();
    copy.into()
  }

  pub fn to_choice_mapping(&self) -> ShapeChoiceMapping {
    self.into()
  }
}
