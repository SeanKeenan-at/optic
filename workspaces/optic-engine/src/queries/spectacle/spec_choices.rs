use crate::commands::ShapeCommand;
use crate::projections::ShapeProjection;
use crate::queries::ShapeQueries;
use crate::shapes::ShapeTrail;
use crate::state::shape::{FieldId, ShapeId, ShapeKind};
use crate::state::SpecIdGenerator;
use serde::Serialize;
use std::collections::BTreeSet;

#[derive(Debug, Serialize, Clone, Eq, PartialEq, Ord, PartialOrd)]
pub enum JsonType {
  String,
  Number,
  Boolean,
  Array,
  Object,
  Null,
  Undefined,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PrimitiveChoice {
  shape_id: ShapeId,
  json_type: JsonType,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ObjectFieldChoice {
  name: String,
  field_id: FieldId,
  shape_id: ShapeId,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ObjectChoice {
  shape_id: ShapeId,
  json_type: JsonType,
  fields: Vec<ObjectFieldChoice>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ArrayChoice {
  json_type: JsonType,
  shape_id: ShapeId,
  item_shape_id: ShapeId,
}

#[derive(Debug, Serialize, Clone)]
#[serde(tag = "type")]
pub enum ShapeChoice {
  Primitive(PrimitiveChoice),
  Object(ObjectChoice),
  Array(ArrayChoice),
  Any,
  Unknown,
}

impl ShapeChoice {
  fn json_type(&self) -> Option<&JsonType> {
    match self {
      ShapeChoice::Primitive(choice) => Some(&choice.json_type),
      ShapeChoice::Object(choice) => Some(&choice.json_type),
      ShapeChoice::Array(choice) => Some(&choice.json_type),
      ShapeChoice::Any => None,
      ShapeChoice::Unknown => None,
    }
  }
}

pub struct ShapeChoiceQueries<'a> {
  shape_queries: ShapeQueries<'a>,
  shape_projection: &'a ShapeProjection,
}

impl<'a> From<&'a ShapeProjection> for ShapeChoiceQueries<'a> {
  fn from(shape_projection: &'a ShapeProjection) -> Self {
    let shape_queries = ShapeQueries::new(shape_projection);

    Self {
      shape_queries,
      shape_projection,
    }
  }
}

impl<'a> ShapeChoiceQueries<'a> {
  pub fn trail_choices(
    &'a self,
    shape_trail: &ShapeTrail,
  ) -> impl Iterator<Item = ShapeChoice> + 'a {
    let trail_choices = self.shape_queries.list_trail_choices(&shape_trail);
    let queries = &self.shape_queries;

    trail_choices
      .into_iter()
      .map(move |choice| match choice.core_shape_kind {
        ShapeKind::ObjectKind => {
          let object_fields = queries.resolve_shape_field_id_and_names(&choice.shape_id);
          let fields = object_fields
            .map(|(field_id, name)| {
              let field_shape_id = queries
                .resolve_field_shape_node(field_id)
                .expect("expected field shape to resolve");

              ObjectFieldChoice {
                field_id: field_id.clone(),
                shape_id: field_shape_id,
                name: name.clone(),
              }
            })
            .collect();

          let output = ObjectChoice {
            shape_id: choice.shape_id.clone(),
            json_type: JsonType::Object,
            fields,
          };
          ShapeChoice::Object(output)
        }
        ShapeKind::ListKind => {
          let shape_parameter_id = &String::from(
            choice
              .core_shape_kind
              .get_parameter_descriptor()
              .expect("expected $list to have a parameter descriptor")
              .shape_parameter_id,
          );
          let list_item_shape_id =
            queries.resolve_parameter_to_shape(&choice.shape_id, shape_parameter_id);
          let output = ArrayChoice {
            shape_id: choice.shape_id.clone(),
            json_type: JsonType::Array,
            item_shape_id: list_item_shape_id,
          };
          ShapeChoice::Array(output)
        }
        ShapeKind::MapKind => unimplemented!(),
        ShapeKind::OneOfKind => unreachable!(),
        ShapeKind::AnyKind => ShapeChoice::Any,
        ShapeKind::UnknownKind => ShapeChoice::Unknown,
        ShapeKind::StringKind => ShapeChoice::Primitive(PrimitiveChoice {
          shape_id: choice.shape_id.clone(),
          json_type: JsonType::String,
        }),
        ShapeKind::NumberKind => ShapeChoice::Primitive(PrimitiveChoice {
          shape_id: choice.shape_id.clone(),
          json_type: JsonType::Number,
        }),
        ShapeKind::BooleanKind => ShapeChoice::Primitive(PrimitiveChoice {
          shape_id: choice.shape_id.clone(),
          json_type: JsonType::Boolean,
        }),
        ShapeKind::NullableKind => ShapeChoice::Primitive(PrimitiveChoice {
          shape_id: choice.shape_id.clone(),
          json_type: JsonType::Null,
        }),
        ShapeKind::IdentifierKind => unimplemented!(),
        ShapeKind::ReferenceKind => unimplemented!(),
        ShapeKind::OptionalKind => ShapeChoice::Primitive(PrimitiveChoice {
          shape_id: choice.shape_id.clone(),
          json_type: JsonType::Undefined,
        }),
      })
  }

  pub fn edit_field_commands(
    &'a self,
    field_id: &FieldId,
    requested_json_types: impl IntoIterator<Item = &'a JsonType> + 'a,
    id_generator: &'a mut impl SpecIdGenerator,
  ) -> Option<impl Iterator<Item = ShapeCommand> + 'a> {
    let field_shape_trail = self.shape_queries.resolve_shape_trail(field_id)?;

    dbg!(&field_shape_trail);

    let requested_kinds =
      requested_json_types
        .into_iter()
        .filter_map(|json_type| match json_type {
          JsonType::Undefined => Some(&ShapeKind::OptionalKind),
          JsonType::Null => Some(&ShapeKind::NullableKind),
          _ => None,
        });

    self
      .shape_queries
      .edit_shape_trail_commands(&field_shape_trail, requested_kinds, id_generator)
  }
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::commands::SpecCommand;
  use crate::events::SpecEvent;
  use crate::projections::SpecProjection;
  use crate::Aggregate;
  use insta::assert_debug_snapshot;
  use serde_json::json;

  #[test]
  pub fn can_generate_edit_field_commands() {
    let events: Vec<SpecEvent> = serde_json::from_value(json!([
      { "ShapeAdded": { "shapeId": "string_shape_1", "baseShapeId": "$string", "name": "", "eventContext": null }},
      { "ShapeAdded": { "shapeId": "object_shape_1", "baseShapeId": "$object", "name": "", "eventContext": null }},
      { "FieldAdded": { "fieldId": "field_1", "shapeId": "object_shape_1", "name": "lastName", "shapeDescriptor": { "FieldShapeFromShape": { "fieldId": "field_1", "shapeId": "string_shape_1"}}, "eventContext": null }},
    ]))
    .expect("should be able to deserialize test events");

    let spec_projection = SpecProjection::from(events);
    let queries = ShapeChoiceQueries::from(spec_projection.shape());

    let mut id_generator = SequentialIdGenerator { next_id: 1093 }; // <3 primes
    let required_json_types = vec![JsonType::Undefined, JsonType::Null];

    let edit_shape_commands = queries
      .edit_field_commands(
        &"field_1".to_owned(),
        &required_json_types,
        &mut id_generator,
      )
      .expect("field should be able to be made required and nullable")
      .map(SpecCommand::from)
      .collect::<Vec<_>>();

    assert_debug_snapshot!(
      "can_generate_edit_field_commands__commands",
      &edit_shape_commands
    );

    let updated_spec = assert_valid_commands(spec_projection.clone(), edit_shape_commands);
    let updated_queries = ShapeChoiceQueries::from(updated_spec.shape());
    let updated_choices = updated_queries
      .trail_choices(&ShapeTrail::new("object_shape_1".to_owned()))
      .collect::<Vec<_>>();

    assert_debug_snapshot!(
      "can_generate_edit_field_commands__updated_choices",
      updated_choices
    );
  }

  fn assert_valid_commands(
    mut spec_projection: SpecProjection,
    commands: impl IntoIterator<Item = SpecCommand>,
  ) -> SpecProjection {
    // let mut spec_projection = SpecProjection::default();
    for command in commands {
      let events = spec_projection
        .execute(command)
        .expect("generated commands must be valid");

      for event in events {
        spec_projection.apply(event)
      }
    }

    spec_projection
  }

  #[derive(Debug, Default)]
  struct SequentialIdGenerator {
    next_id: u32,
  }
  impl SpecIdGenerator for SequentialIdGenerator {
    fn generate_id(&mut self, prefix: &str) -> String {
      self.next_id += 1;
      format!("{}{}", prefix, self.next_id.to_string())
    }
  }
}
