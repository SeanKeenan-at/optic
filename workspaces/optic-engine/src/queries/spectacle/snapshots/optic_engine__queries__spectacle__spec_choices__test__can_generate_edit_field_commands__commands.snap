---
source: workspaces/optic-engine/src/queries/spectacle/spec_choices.rs
expression: "&edit_shape_commands"
---
[
    ShapeCommand(
        AddShape(
            AddShape {
                shape_id: "shape_1094",
                base_shape_id: "$nullable",
                name: "",
            },
        ),
    ),
    ShapeCommand(
        SetParameterShape(
            SetParameterShape {
                shape_descriptor: ProviderInShape(
                    ProviderInShape {
                        shape_id: "shape_1094",
                        provider_descriptor: ShapeProvider(
                            ShapeProvider {
                                shape_id: "string_shape_1",
                            },
                        ),
                        consuming_parameter_id: "$nullableInner",
                    },
                ),
            },
        ),
    ),
    ShapeCommand(
        AddShape(
            AddShape {
                shape_id: "shape_1095",
                base_shape_id: "$optional",
                name: "",
            },
        ),
    ),
    ShapeCommand(
        SetParameterShape(
            SetParameterShape {
                shape_descriptor: ProviderInShape(
                    ProviderInShape {
                        shape_id: "shape_1095",
                        provider_descriptor: ShapeProvider(
                            ShapeProvider {
                                shape_id: "shape_1094",
                            },
                        ),
                        consuming_parameter_id: "$optionalInner",
                    },
                ),
            },
        ),
    ),
    ShapeCommand(
        SetFieldShape(
            SetFieldShape {
                shape_descriptor: FieldShapeFromShape(
                    FieldShapeFromShape {
                        field_id: "field_1",
                        shape_id: "shape_1095",
                    },
                ),
            },
        ),
    ),
]
